### PR TITLE
Build config/update to gradle 8.7

### PR DESCRIPTION
**Pull request by dkrasnoff**
In this PR:

- Bumped gradle version to 8.7
- Removed `atomicfu-native` module - old unnecessary module
- Remove `disable_native_targets` flag - was used on CI before, now we don't have such flag
- Converted root `build.gradle` to `build.gradle.kts`. And removed block with printing `META-INF/MANIFEST.MF` to log during build, because it does not used anywhere. 
- Converted last shared build script from `gradle` dir to convetions plugin (`kotlin-multiplatform-publish-conventions.gradle.kts`)

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,6 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.3-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.7-bin.zip
```
Comment (adam-enko) [inline gradle/wrapper/gradle-wrapper.properties:3]: It's not enough to just update the Gradle version here - you need to run `gradle wrapper` to update the wrapper files.

https://docs.gradle.org/8.7/userguide/gradle_wrapper.html#sec:upgrading_wrapper

**Review by adam-enko — CHANGES_REQUESTED** [review]

File: atomicfu/src/nativeInterop/cinterop/interop.def
Comment (adam-enko) [inline atomicfu/src/nativeInterop/cinterop/interop.def:1]: It's strange seeing code changes in this PR. Isn't this PR just supposed to update the Gradle version and modernise the build config?

If code changes are required, can they be split into a separate PR? It'd help keep the changes organised, and if there's a bug in either of the PRs then we can more easily identify whether it was caused by a change in the build config or the source code.

File: atomicfu-maven-plugin/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+val mavenUserHome = System.getProperty("maven.user.home")
+val mavenRepoLocal = System.getProperty("maven.repo.local")
+
+val generatePomFileForMavenPublication by tasks.getting(GenerateMavenPom::class)
+
+// runs the plugin description generator
+val generatePluginDescriptor by tasks.registering(Exec::class) {
+    dependsOn(generatePomFileForMavenPublication, project(":atomicfu-transformer").tasks.named("publishToMavenLocal"))
```
Comment (adam-enko) [inline atomicfu-maven-plugin/build.gradle.kts:32]:
```suggestion
    dependsOn(generatePomFileForMavenPublication)
    dependsOn(project(":atomicfu-transformer").tasks.named("publishToMavenLocal"))
```

formatting nitpick: put the `dependsOn()` on separate lines, so it's easier to read.

File: atomicfu-maven-plugin/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+publishing.publications {
+    named<MavenPublication>("maven").configure {
+        pom.configureMavenPluginPomAttributes(project, outputDir.get().asFile.path)
+        pom.packaging = "maven-plugin"
+    }
+}
+
+val mavenUserHome = System.getProperty("maven.user.home")
+val mavenRepoLocal = System.getProperty("maven.repo.local")
```
Comment (adam-enko) [inline atomicfu-maven-plugin/build.gradle.kts:26]:
```suggestion
val mavenUserHome: String? = System.getProperty("maven.user.home")
val mavenRepoLocal: String? = System.getProperty("maven.repo.local")
```

Let's specify the types for clarity and to avoid the "Declaration has type inferred from a platform call..." warning.

File: build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import org.jetbrains.kotlin.gradle.targets.js.nodejs.NodeJsRootExtension
+import org.jetbrains.kotlin.gradle.targets.js.npm.tasks.KotlinNpmInstallTask
+
+plugins {
+    alias(libs.plugins.kotlinx.binaryCompatibilityValidator)
+}
+
+val deploy by tasks.creating() {
```
Comment (adam-enko) [inline build.gradle.kts:8]:
```suggestion
val deploy by tasks.creating {
```

unnecessary brackets

File: build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    dependsOn(getTasksByName("publish", true))
+    dependsOn(getTasksByName("publishNpm", true))
+}
+
+// Right now it is used for switching nodejs version which is supports generated wasm bytecode
```
Comment (adam-enko) [inline build.gradle.kts:13]:
```suggestion
// Right now it is used for switching nodejs version which is supports generated wasm bytecode - Remove after updating to Kotlin 2.0
```

File: build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+extensions.findByType(NodeJsRootExtension::class.java)?.let {
+    // canary nodejs that supports recent Wasm GC changes
+    it.nodeVersion = "21.0.0-v8-canary202309167e82ab1fa2"
+    it.nodeDownloadBaseUrl = "https://nodejs.org/download/v8-canary"
+}
+
+// We need to ignore unsupported engines (i.e. canary) for npm
```
Comment (adam-enko) [inline build.gradle.kts:20]:
```suggestion
// We need to ignore unsupported engines (i.e. canary) for npm  - Remove after updating to Kotlin 2.0
```

File: build.gradle.kts (hunk 5 above)
Comment (adam-enko) [reply build.gradle.kts:13]: Actually, I don't think this section will do anything. KGP isn't applied to the root project, so `NodeJsRootExtension` won't be present until later... perhaps it can be removed?

File: build.gradle.kts (hunk 6 above)
Comment (lppedd) [reply build.gradle.kts:20]: @adam-enko I'm not sure what the plan is for 2.0, but I'd release an intermediate version with 1.9.24, which allows to get rid of canary anyway.

Comment (adam-enko) [reply build.gradle.kts:20]: Thanks for the tip @lppedd! It certainly would be nice to be able to remove this npm-nightly config.

File: atomicfu/src/nativeInterop/cinterop/interop.def
Comment (dkrasnoff) [reply atomicfu/src/nativeInterop/cinterop/interop.def:1]: oh, I didn't change this file, just mistaken `master` branch instead of `develop`. 
Changed it to right branch, thanks!

File: gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (dkrasnoff) [reply gradle/wrapper/gradle-wrapper.properties:3]: fixed

File: atomicfu-maven-plugin/build.gradle.kts (hunk 3 above)
Comment (dkrasnoff) [reply atomicfu-maven-plugin/build.gradle.kts]: fixed

File: atomicfu-maven-plugin/build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            "--errors",
+            "--batch-mode",
+            "--file", generatePomFileForMavenPublication.destination.toString(),
+            "org.apache.maven.plugins:maven-plugin-plugin:3.5.1:descriptor"
+        )
+    )
+    commandLine = args
+    doLast {
+        val descriptorFile = pluginDescriptorFile.get().asFile
+        assert(descriptorFile.exists()) { "$descriptorFile: was not generated" }
```
[the remark this answers is not in the document]
Comment (dkrasnoff) [reply atomicfu-maven-plugin/build.gradle.kts]: fixed

File: build.gradle.kts (hunk 4 above)
Comment (dkrasnoff) [reply build.gradle.kts]: fixed

File: build.gradle.kts (hunk 5 above)
Comment (dkrasnoff) [reply build.gradle.kts]: Locally it succeeded. I pushed the commit with removing to check on CI.

Comment (dkrasnoff) [reply build.gradle.kts]: Hmm, on CI [we failed ](https://teamcity.jetbrains.com/buildConfiguration/KotlinTools_KotlinxAtomicfu_BuildAggregatedLinuxWindows/4665709?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true), thus I will not remove it for now

**Comment by mvicsokolova** [comment]
Thank you! Looks good 🙂

**Review by mvicsokolova — APPROVED** [review]

File: build.gradle.kts (hunk 5 above)
Comment (adam-enko) [reply build.gradle.kts]: Hmm strange! Let's just update the comment and tidy up the K/JS config in a later PR.

**Review by adam-enko — APPROVED** [review]
Nice work!

File: build.gradle.kts (hunk 6 above)
Comment (adam-enko) [reply build.gradle.kts:20]: Discussed with @dkrasnoff  - we'll tidy up the K/JS config in a later PR.
